lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -1183,6 +1209,59 @@ ConstExprFunctionState::computeSingleStoreAddressValue(SILValue addr) {
       return evaluator.getUnknown(addr, UnknownReason::Default);
     }
 
+    if (auto *teai = dyn_cast<TupleElementAddrInst>(user)) {
+      // Try finding a writer among its users. For example:
+      //   %132 = tuple_element_addr %131 : $*(Int32, Int32, Int32, Int32), 0
+      // If %132 is written to as a const value in one of its users, the call
+      // below will return a constant. An example writer can be:
+      //   copy_addr %111 to [initialization] %132
+      // Otherwise we continue to search for a writer.
+
+      auto *use = teai->getSingleUse();
+      if (!use)
+        continue;
+
+      // In this case we want to find a user of teai that writes to that tuple
+      // elt addr. We cannot call computeSingleStoreAddressValue(), since
+      // that'll involve evaluating the tuple operand (first operand) of teai
+      // first, denoted as X (an inst that creates the tuple), and that in turn
+      // will go through all uses of X, include teai, entering a cycle.
+
+      // Instead we call the lighter-weight method
+      // computeSingleStoreAddressValueHelper() to avoid forming the cycle.
+      auto tupleEltAddrValue = computeSingleStoreAddressValueHelper(teai);
+      if (tupleEltAddrValue.isConstant()) {
+        // If the tuple elt is indeed a const, we write it into (the appropriate
+        // aggregate element slot of the) `memoryObject`, which is the const
+        // value for the entire tuple.
+        SmallVector<unsigned, 4> accessPath;
+        auto *tupleEltMemoryObject =
+            tupleEltAddrValue.getAddressValue(accessPath);
+        assert(accessPath.empty());
+        auto tupleEltValue = tupleEltMemoryObject->getValue();
+        assert(tupleEltValue.isConstant());
+
+        auto objectVal = memoryObject->getValue();
+        auto objectType = memoryObject->getType();
+        auto index = teai->getFieldNo();
+        bool failed = updateIndexedElement(objectVal, /*accessPath*/ {index},
+                                           tupleEltValue, objectType,
+                                           evaluator.getAllocator());
+        if (failed)
+          return evaluator.getUnknown(addr, UnknownReason::Default);
+        memoryObject->setValue(objectVal);
+        // If all aggregate elements are const, we have successfully
+        // const-evaluated the entire tuple!
+        if (llvm::all_of(objectVal.getAggregateValue(),
+                         [](SymbolicValue v) { return v.isConstant(); }))
+          break;
+        // Otherwise we keep going, hoping to find more users of the tuple inst
+        // that write into other tuple elements.
```

Comment:
Is this the right thing to do? Can't we have a use of the tuple before all values are initialized?

Reply:
Good point on dealing with multiple writers. I added a `writeOnlyOnce` param, to make sure we won't have multiple writers to the same agg element.

Not sure what'd be a swift test case that can trigger "multiple writers". From our discussion, please feel free to follow up on this. :)

Regarding "use of a tuple before all values are initialized", I believe SILGen cannot generate such code.

Reply:
Thanks for the clarifications. 

Actually, the existing implementation of  computeSingleStoreAddressValue returns unknown in such cases. See the case for StoreInst:
`    if (auto *si = dyn_cast<StoreInst>(user)) {
       if (use->getOperandNumber() == 1) {

         // If we have already found a value for this stack slot then we're done:                                                                                                                                                                                               
         // we don't support multiple assignment.                                                                                                                                                                                                                               
         if (memoryObject->getValue().getKind() != SymbolicValue::UninitMemory)
           return evaluator.getUnknown(addr, UnknownReason::Default);
        ...
`
Should we have the same behavior here?

Reply:
Yes. the code below should do it
```c++
if (failed)
    return evaluator.getUnknown(addr, UnknownReason::Default);
```